server/src/main/java/org/opensearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -149,6 +168,202 @@ public TransportShardBulkAction(
+         **/
+        private final String targetAllocationID;
+        private final long primaryTerm;
+        private final ShardId shardId;
+
+        public PrimaryTermValidationRequest(String targetAllocationID, long primaryTerm, ShardId shardId) {
+            this.targetAllocationID = Objects.requireNonNull(targetAllocationID);
+            this.primaryTerm = primaryTerm;
+            this.shardId = Objects.requireNonNull(shardId);
+        }
+
+        public PrimaryTermValidationRequest(StreamInput in) throws IOException {
+            super(in);
+            targetAllocationID = in.readString();
+            primaryTerm = in.readVLong();
+            shardId = new ShardId(in);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeString(targetAllocationID);
+            out.writeVLong(primaryTerm);
+            shardId.writeTo(out);
+        }
+
+        @Override
+        public Task createTask(long id, String type, String action, TaskId parentTaskId, Map<String, String> headers) {
+            return new ReplicationTask(id, type, action, getDescription(), parentTaskId, headers);
+        }
+
+        public String getTargetAllocationID() {
+            return targetAllocationID;
+        }
+
+        public long getPrimaryTerm() {
+            return primaryTerm;
+        }
+
+        public ShardId getShardId() {
+            return shardId;
+        }
+
+        @Override
+        public String getDescription() {
+            return toString();
+        }
+
+        @Override
+        public String toString() {
+            return "PrimaryTermValidationRequest ["
+                + shardId
+                + "] for targetAllocationID ["
+                + targetAllocationID
+                + "] with primaryTerm ["
+                + primaryTerm
+                + "]";
+        }
+    }
+
+    @Override
+    protected ReplicationOperation.Replicas<BulkShardRequest> newReplicasProxy() {
+        return new PrimaryTermValidationProxy();
+    }
+
+    private final class PrimaryTermValidationProxy extends WriteActionReplicasProxy {
+        @Override
+        public void performOn(
+            ShardRouting replica,
+            BulkShardRequest request,
+            long primaryTerm,
+            long globalCheckpoint,
+            long maxSeqNoOfUpdatesOrDeletes,
+            ActionListener<ReplicationOperation.ReplicaResponse> listener,
+            ReplicationMode replicationMode
+        ) {
+            assert replicationMode == ReplicationMode.FULL_REPLICATION || replicationMode == ReplicationMode.PRIMARY_TERM_VALIDATION;
+
+            if (replicationMode == ReplicationMode.FULL_REPLICATION) {
+                super.performOn(replica, request, primaryTerm, globalCheckpoint, maxSeqNoOfUpdatesOrDeletes, listener, replicationMode);
+                return;
+            }
```

Comment:
Why do we want a `PrimaryTermValidationProxy ` to handle full replication?

Reply:
This is to handle primary term validation which is specific to TransportShardBulkAction. In case of full replication, we let the parent performOn handle the replication call (which is doing replication). For primary term validation, we make the primary term validation request and send to the primary term validation action.